### PR TITLE
Apply thread-local cache to InitLayer and MergingLayer

### PR DESCRIPTION
**Pull request by Gegy**
Previously, we were only applying this to `ParentedLayer`, which would eliminate the majority of potential race conditions. This would make any race condition issue practically impossible in a vanilla context, but when adding more threads, it becomes much more likely to occur through the alternative `InitLayer` and `MergingLayer`, which were missed by accident in the original PR.

### CODE REVIEW

**Review by jaskarth — APPROVED** [review]

**Comment by 2No2Name** [comment]
Looking at the code now. Are you saying the vanilla code has race conditions?

File: src/main/java/me/jellysquid/mods/lithium/common/world/layer/CachedLocalLayerFactory.java [hunk 1]
```diff
@@ -1,42 +1,56 @@
...
+            }, sampler1, sampler2);
+        });
+    }
 
-        return context.createSampler((x, z) -> {
-            context.initSeed(x, z);
-            return layer.sample(context, parentSampler, x, z);
-        }, parentSampler);
+    private static <R extends LayerSampler> LayerFactory<R> createMemoized(LayerFactory<R> factory) {
+        ThreadLocal<R> cached = new ThreadLocal<>();
```
Comment (2No2Name) [inline src/main/java/me/jellysquid/mods/lithium/common/world/layer/CachedLocalLayerFactory.java]: Is this different from `return ThreadLocal.withInitial(factory::make)::get;`?

**Review by 2No2Name — APPROVED** [review]
In general the code looks good to me. I wonder how you chose the cache capacities, as they are different from vanilla, but you are not sharing the cache between threads.

File: src/main/java/me/jellysquid/mods/lithium/common/world/layer/CachedLocalLayerFactory.java (hunk 1 above)
Comment (Gegy) [reply src/main/java/me/jellysquid/mods/lithium/common/world/layer/CachedLocalLayerFactory.java]: That should work

**Comment by Gegy** [comment]
I don't remember exactly how we ended up with these cache sizes- I recall there was the consideration of how many threads there are, but also the reduced memory overhead with this implementation in comparison to vanilla.

**Pull request merged**